src/flow/src/batching_mode/engine.rs

```diff
@@ -0,0 +1,1151 @@
+
+        let elapsed = instant.elapsed();
+        if let Ok(res1) = &res {
+            debug!(
+                "Flow {flow_id} executed, result: {res1:?}, elapsed: {:?}",
+                elapsed
+            );
+        } else if let Err(res) = &res {
+            warn!(
+                "Failed to execute Flow {flow_id} on frontend {}, result: {res:?}, elapsed: {:?} with query: {}",
+                peer_addr, elapsed, &sql
+            );
+        }
+
+        // record slow query
+        if elapsed >= SLOW_QUERY_THRESHOLD {
+            warn!(
+                "Flow {flow_id} on frontend {} executed for {:?} before complete, query: {}",
+                peer_addr, elapsed, &sql
+            );
+            METRIC_FLOW_BATCHING_ENGINE_SLOW_QUERY
+                .with_label_values(&[flow_id.to_string().as_str(), sql, &peer_addr])
+                .observe(elapsed.as_secs_f64());
+        }
+
+        self.state
+            .write()
+            .await
+            .after_query_exec(elapsed, res.is_ok());
+
+        let res = res.context(InvalidRequestSnafu {
+            context: format!(
+                "Failed to execute query for flow={}: \'{}\'",
+                self.flow_id, &sql
+            ),
+        })?;
+
+        Ok(Some((res, elapsed)))
+    }
+
+    /// start executing query in a loop, break when receive shutdown signal
+    pub async fn start_executing_loop(
+        &self,
+        engine: QueryEngineRef,
+        frontend_client: Arc<FrontendClient>,
+    ) -> Result<(), Error> {
+        loop {
+            let Some(_) = self.gen_exec_once(&engine, &frontend_client).await? else {
+                debug!(
+                    "Flow id = {:?} found no new data, sleep for {:?} then continue",
+                    self.flow_id, MIN_REFRESH_DURATION
+                );
+                tokio::time::sleep(MIN_REFRESH_DURATION).await;
+                continue;
+            };
+
+            let sleep_until = {
+                let mut state = self.state.write().await;
+                match state.shutdown_rx.try_recv() {
+                    Ok(()) => break Ok(()),
+                    Err(TryRecvError::Closed) => {
+                        warn!("Unexpected shutdown flow {}, shutdown anyway", self.flow_id);
+                        break Ok(());
+                    }
+                    Err(TryRecvError::Empty) => (),
+                }
+                state.get_next_start_query_time(None)
+            };
+            tokio::time::sleep_until(sleep_until).await;
+        }
+    }
+
+    /// Generate the create table SQL
+    ///
+    /// the auto created table will automatically added a `update_at` Milliseconds DEFAULT now() column in the end
+    /// (for compatibility with flow streaming mode)
+    ///
+    /// and it will use first timestamp column as time index, all other columns will be added as normal columns and nullable
+    async fn gen_create_table_sql(&self, engine: QueryEngineRef) -> Result<String, Error> {
```

Comment:
SQL injection warning! Why not create the plan or statement instead?

Reply:
> SQL injection warning! Why not create the plan or statement instead?

will restore frontend's grpc logical plan handle path for this in this PR, previously it's deemed unnecessary since user of grpc protocol shouldn't send logical plan